File: ch5/clip6/boj1932/main.py
n = int(input())
a = [[0] * (n + 1) for _ in range(n + 1)]
d = [[0] * (n + 1) for _ in range(n + 1)]

for i in range(1, n + 1):
    line = list(map(int, input().split()))
    for j in range(1, i + 1):
        a[i][j] = line[j - 1]

d[1][1] = a[1][1]

for i in range(2, n + 1):
    for j in range(1, i + 1):
        d[i][j] = max(d[i - 1][j - 1], d[i - 1][j]) + a[i][j]

for i in range(1, n):
    for j in range(1, i + 1):
        d[i + 1][j] = max(d[i + 1][j], d[i][j] + a[i + 1][j])
        d[i + 1][j + 1] = max(d[i + 1][j + 1], d[i][j] + a[i + 1][j + 1])

ans = 0
for i in range(1, n + 1):
    ans = max(ans, d[n][i])

print(ans)

File: ch5/clip6/boj1932/Main.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n  = sc.nextInt();
        int[][] d = new int[n + 1][n + 1];
        int[][] a = new int[n + 1][n + 1];
        for(int i = 1; i <= n; i++) {
            for(int j = 1; j <= i; j++) {
                a[i][j] = sc.nextInt();
            }
        }
        d[1][1] = a[1][1];
        for(int i = 2; i <= n; i++) {
            for(int j = 1; j <= i; j++) {
                d[i][j] = Math.max(d[i - 1][j - 1], d[i - 1][j]) + a[i][j];
            }
        }
        int ans = 0;
        for(int i = 1; i <= n; i++) {
            ans = Math.max(ans, d[n][i]);
        }
        System.out.println(ans);
    }
}
